### DISCUSSION TITLE
Reg mmap profiling

### DISCUSSION THREAD

**Discussion by vinayhudli** [async-profiler/async-profiler#453] [Q&A] [answered]
Hi,
Scenario : I profiled my application using async-profiler for mmap calls for about 15 minutes to debug native memory leak. In the flamegraph o/p I see only 3 samples for the java code and the rest is not the application call.
Problem : Does this mean that mmap was called only 3 times in 15 minutes ? This cannot be true because the method responsible for calling mmap is called atleast 10 times per sec. Is there any way to profile for every x kb of mmap allocation ?

Attaching sample flamegraph for clarity.
<img width="1660" alt="part of code that should be called lot of times" src="https://user-images.githubusercontent.com/1204958/131457841-dbd3c73b-b805-47d9-bad2-77a74f6f0aa7.png">

<img width="1666" alt="whole flamegraph" src="https://user-images.githubusercontent.com/1204958/131457971-2184b47d-0e16-4fb6-996d-4e0a3e56ec0f.png">

**Comment by vinayhudli** [comment]
Thanks for the response. I'm mostly interested in "mmap" calls because I
suspect there is native memory leak and based on some searching I think
mmap is the place to look at. I'm using below mentioned command to profile
for mmap allocations:
./profile.sh -d 900 -e mmap -f mmap.html 1
And the call stack that has only 3 samples should have been called way more
than 3 over a period of 900 s duration.
Furthermore, I think frequency is only for cpu profiling ? Please correct
me if I'm wrong.

Thanks,
Vinay

**Comment by krzysztofslusarski** [comment]
For native memory leak I suggest you tu use ```page-faults``` event. ```mmap``` doesn't mean that your RSS is growing, because after ```mmap``` your application can do ```munmap```. From my experience ```page-faults``` is the best way to diagnose native memory leak, because it shows you where your RSS grows.

**Reply by vinayhudli** [reply]
@krzysztofslusarski I tried page faults in the past, attaching that graph here. The stack trace is mostly of method that downloads file from on-prem storage and uploads to cloud storage, these files are small in size (< 5 mb) so I read bytes from the file and upload to cloud storage 
<img width="1671" alt="Screen Shot 2021-08-31 at 3 54 50 PM" src="https://user-images.githubusercontent.com/1204958/131586436-2793fe43-0a0e-424c-aa40-648f7fd83eb8.png">
.

**Comment by apangin** [comment, marked as answer]
> Does this mean that mmap was called only 3 times in 15 minutes ?

Yes, this is most likely true.
`-e mmap` traces all calls to libc `mmap` function.

Strictly speaking, there is a way to invoke `mmap` directly via syscall, bypassing the libc function. However, this is an exotic scenario, applications typically don't do that.

Anyway, there is another way to trace all `mmap`s, including the direct syscalls: `-e syscalls::sys_exit_mmap`

> This cannot be true because the method responsible for calling mmap is called atleast 10 times per sec.

You probably mean a method that allocates native memory? Not every native allocation result in `mmap` call. I'd say, in a normal situation (when there is no memory leak), almost no allocation should end up with `mmap`. The entire idea of profiling `mmap` is to catch only those allocations that result in *growth* of the process RSS. So, if the profiler shows only a few `mmap` calls, this probably means you have no native memory leak, or the leak is very very slow.

As to native memory leaks, I have a good news for you.
Async-profiler recently got the new mode `-e nativemem` specifically for finding native memory leaks. It is currently an experimental feature, so feedback is welcome. It is available in [`malloc`](https://github.com/jvm-profiling-tools/async-profiler/tree/malloc) branch.

See [README](https://github.com/jvm-profiling-tools/async-profiler/tree/malloc#finding-native-memory-leaks) for details.

**Reply by vinayhudli** [reply]
```
You probably mean a method that allocates native memory?  
```
My app doesn't allocate native memory directly but uses a library to upload to google cloud storage that seems to be eventually calling mmap. 

```
So, if the profiler shows only a few mmap calls, this probably means you have no native memory leak, or the leak is very very slow
```
So the issue is RSS keeps growing when there is work to be done and when there are no more requests it just stays at that level and doesn't release memory. When the new requests start coming in it again grows from that point. I have seen it release it very little memory sometimes to make way for new requests probably.

I can try the new mode, thanks

**Reply by apangin** [reply]
Sound like one of the typical problems:
 - cached direct buffers;
 - unclosed I/O streams or deflaters;
 - malloc arena fragmentation.

Things to try:
 - limit `-XX:MaxDirectMemorySize` and make sure `-XX:+DisableExplicitGC` is not set;
 - change the standard allocator to jemalloc;
 - watch https://www.youtube.com/watch?v=c755fFv1Rnk
 - and the new mode of async-profiler, of course :)

**Comment by krzysztofslusarski** [comment]
Wow, that mode looks nice. Thank you @apangin.
